cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll

```diff
@@ -0,0 +1,326 @@
+/**
+ * Provides implementation classes modeling C++ iterators, including
+ * `std::iterator`, `std::iterator_traits`, and types meeting the
+ * `LegacyIterator` named requirement. See `semmle.code.cpp.models.Models` for
+ * usage information.
+ */
+
+import cpp
+import semmle.code.cpp.models.interfaces.Taint
+import semmle.code.cpp.models.interfaces.DataFlow
+
+/**
+ * An instantiation of the `std::iterator_traits` template.
+ */
+class IteratorTraits extends Class {
+  IteratorTraits() {
+    this.hasQualifiedName("std", "iterator_traits") and
+    not this instanceof TemplateClass and
+    exists(TypedefType t |
+      this.getAMember() = t and
+      t.getName() = "iterator_category"
+    )
+  }
+
+  Type getIteratorType() { result = this.getTemplateArgument(0) }
+}
+
+/**
+ * A type which has the typedefs expected for an iterator.
+ */
+class IteratorByTypedefs extends Class {
+  IteratorByTypedefs() {
+    this.getAMember().(TypedefType).hasName("difference_type") and
+    this.getAMember().(TypedefType).hasName("value_type") and
+    this.getAMember().(TypedefType).hasName("pointer") and
+    this.getAMember().(TypedefType).hasName("reference") and
+    this.getAMember().(TypedefType).hasName("iterator_category") and
+    not this.hasQualifiedName("std", "iterator_traits")
+  }
+}
+
+/**
+ * The `std::iterator` class.
+ */
+class StdIterator extends Class {
+  StdIterator() { this.hasQualifiedName("std", "iterator") }
+}
+
+/**
+ * A type which can be used as an iterator
+ */
+class Iterator extends Type {
+  Iterator() {
+    this instanceof IteratorByTypedefs or
+    exists(IteratorTraits it | it.getIteratorType() = this) or
+    this instanceof StdIterator
+  }
+}
+
+/**
+ * A non-member prefix `operator*` function for an iterator type.
+ */
+class IteratorPointerDereferenceOperator extends Operator, TaintFunction {
+  IteratorPointerDereferenceOperator() {
+    this.hasName("operator*") and
+    this.getACallToThisFunction().getArgument(0).getFullyConverted().getUnderlyingType() instanceof
+      Iterator
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    input.isParameter(0) and
+    output.isReturnValue()
+  }
+}
+
+/**
+ * A non-member `operator++` or `operator--` function for an iterator type.
+ */
+class IteratorCrementOperator extends Operator, DataFlowFunction {
+  IteratorCrementOperator() {
+    (
+      this.hasName("operator++") or
+      this.hasName("operator--")
+    ) and
+    this
+        .getACallToThisFunction()
+        .getArgument(0)
+        .getFullyConverted()
+        .getUnderlyingType()
+        .(ReferenceType)
+        .getBaseType() instanceof Iterator
+  }
+
+  override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
+    input.isParameter(0) and
+    output.isReturnValue()
+  }
+}
+
+/**
+ * A non-member `operator->` function for an iterator type.
```

Comment:
AFAICT from https://en.cppreference.com/w/cpp/language/operator_member_access, there is no non-member version of this operator.